esrally/telemetry.py

```diff
@@ -568,6 +568,112 @@ def record(self):
+            # we don't require starting recorders precisely at the same time
+            sampler.start()
+
+    def on_benchmark_stop(self):
+        if self.samplers:
+            for sampler in self.samplers:
+                sampler.finish()
+
+
+class ShardStatsRecorder:
+    """
+    Collects and pushes shard stats for the specified cluster to the metric store.
+    """
+
+    def __init__(self, cluster_name, client, metrics_store, sample_interval):
+        """
+        :param cluster_name: The cluster_name that the client connects to, as specified in target.hosts.
+        :param client: The Elasticsearch client for this cluster.
+        :param metrics_store: The configured metrics store we write to.
+        :param sample_interval: integer controlling the interval, in seconds, between collecting samples.
+        """
+
+        self.cluster_name = cluster_name
+        self.client = client
+        self.metrics_store = metrics_store
+        self.sample_interval = sample_interval
+        self.logger = logging.getLogger(__name__)
+
+    def __str__(self):
+        return "shard stats"
+
+    def record(self):
+        """
+        Collect node-stats?level=shards and push to metrics store.
+        """
+        # pylint: disable=import-outside-toplevel
+        import elasticsearch
+        try:
+            sample = self.client.nodes.stats(metric="_all", level="shards")
+        except elasticsearch.TransportError:
+            msg = f"A transport error occurred while collecting shard stats on cluster [{self.cluster_name}]"
+            self.logger.exception(msg)
+            raise exceptions.RallyError(msg)
+
+        shard_metadata = {
+                "cluster": self.cluster_name
+            }
+
+        for node_stats in sample["nodes"].values():
+            node_name = node_stats["name"]
+            collected_node_stats = collections.OrderedDict()
+            collected_node_stats["name"] = "shard-stats"
+            shard_stats = node_stats["indices"].get("shards")
+
+            for index_name, stats in shard_stats.items():
+                for curr_shard in stats:
+                    for shard_id, curr_stats in curr_shard.items():
+                        doc = {
+                            "name": "shard-stats",
+                            "shard-id": shard_id,
+                            "index": index_name,
+                            "primary": curr_stats.get("routing", {}).get("primary"),
+                            "docs": curr_stats.get("docs", {}).get("count", -1),
+                            "store": curr_stats.get("store", {}).get("size_in_bytes", -1),
+                            "segments_count":  curr_stats.get("segments", {}).get("count", -1),
+                            "node": node_name
+                        }
+                        self.metrics_store.put_doc(doc, level=MetaInfoScope.cluster, meta_data=shard_metadata)
+                        print(f"shards {doc}")
```

Comment:
Leftover?

Reply:
oops, thank you for catching

Reply:
Can you please remove the print statement?